src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs

```diff
@@ -0,0 +1,101 @@
+using Microsoft.Azure.ServiceBus.Core;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Arcus.Messaging.Pumps.ServiceBus
+{
+    public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessage, AzureServiceBusMessageContext>
+    {
+        /// <summary>
+        ///     Constructor
+        /// </summary>
+        /// <param name="configuration">Configuration of the application</param>
+        /// <param name="logger">Logger to write telemetry to</param>
+        protected AzureServiceBusMessagePump(IConfiguration configuration, ILogger logger)
+            : base(configuration, logger)
+        {
+        }
+
+        /// <summary>
+        ///     Path of the entity to process
+        /// </summary>
+        public string EntityPath { get; private set; }
+
+        /// <summary>
+        ///     Service Bus namespace that contains the entity
+        /// </summary>
+        public string Namespace { get; private set; }
+
+        /// <inheritdoc/>
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            Logger.LogInformation("Creating message pump");
+            var messageReceiver = CreateMessageReceiver();
+            Logger.LogInformation("Starting message pump on entity path {EntityPath} in namespace {Namespace}", EntityPath, Namespace);
+
+            // TODO: Message pump options to not delete for example
+            var messageHandlerOptions = new MessageHandlerOptions(HandleReceivedException);
+            messageReceiver.RegisterMessageHandler(HandleMessage, messageHandlerOptions);
+            Logger.LogInformation("Message pump started");
+
+            await UntilCancelledAsync(stoppingToken);
+
+            Logger.LogInformation("Closing message pump");
+            await messageReceiver.CloseAsync();
+            Logger.LogInformation("Message pump closed : {Time}", DateTimeOffset.UtcNow);
+        }
+
+        private MessageReceiver CreateMessageReceiver()
+        {
+            var connectionString = Configuration.GetValue<string>("ARCUS_SERVICEBUS_QUEUE_CONNECTIONSTRING");
+
+            var serviceBusConnectionStringBuilder = new ServiceBusConnectionStringBuilder(connectionString);
+
+            var messageReceiver = new MessageReceiver(serviceBusConnectionStringBuilder);
+            
+            EntityPath = serviceBusConnectionStringBuilder.EntityPath;
+            Namespace = messageReceiver.ServiceBusConnection.Endpoint.Host;
+
+            return messageReceiver;
+        }
+
+        private async Task HandleReceivedException(ExceptionReceivedEventArgs exceptionEvent)
+        {
+            await HandleReceiveExceptionAsync(exceptionEvent.Exception);
+        }
+
+        private async Task HandleMessage(Message message, CancellationToken cancellationToken)
+        {
+            var correlationInfo = new MessageCorrelationInfo(message.GetTransactionId(), message.CorrelationId);
```

Comment:
we should really check if the message is null or zero bytes

Reply:
Good point, will change

Reply:
I've added a check but I'd assume JsonConvert is smart enough to handle 0 bytes, no? 🤔

Reply:
I'm not sure if we should throw an exception when the message is null.
"normally" that will not happen, but I've seen it happen in IntegrationCloud, when the receiver was stopped.
So instead of throwing, we could just return > no abandoning or retries needed.

Which also makes me think about what you will do with exceptions.... but that's for future updates I assume.

Reply:
I also don't think it will happen, but makes sense. So an if to return if null is enough for you.

Exceptions will be handled by the configured exception handler (`HandleReceiveExceptionAsync`) which will log the error but you can also customize it by overriding.

Reply:
If to return NULL > yes

regarding the exceptions, I was thinking if we should abandon the message immediately when the message is bad, or we just let servicebus deliverycount handle it on ALL exceptions, but this also depends on how it's confgured (peek/complete or receiveDelete).

It might also be a good idea to throw typed (custom) exceptions so appropriate actions can be taken depending on the exception (maybe even include the original message in the exception).

Reply:
Actually that's a good point - Remember seeing the same but what we did at customer was ignore inbound messages as we're in shutting-down-state.

> If to return NULL > yes
> 
> regarding the exceptions, I was thinking if we should abandon the message immediately when the message is bad, or we just let servicebus deliverycount handle it on ALL exceptions, but this also depends on how it's confgured (peek/complete or receiveDelete).

This is indeed my concern.

> It might also be a good idea to throw typed (custom) exceptions so appropriate actions can be taken depending on the exception (maybe even include the original message in the exception).

Not sure how that would look like then, `ReceivedNullMessageException` or what do you mean? Because here we don't have any exception message or do you mean the message that was received?

Since this is the initial start I think it's best to start small and not overengineer, we can still add that later on. Make sense to you?

Reply:
My proposal for that is flag pump in shutdown on close.

After that, every new message will start with following:
```csharp
if (IsShuttingDown)
            {
                if (_messageHandlerOptions.AutoComplete == false)
                {
                    await _messageReceiver.AbandonAsync(message.SystemProperties.LockToken);
                }
                else
                {
// TODO: New exception to be thrown, such as HostShuttingDownException();
                }
            }
```

A similar approach can be done for null messages but with an exception such as `InvalidMessageReceivedException` or so.

Thoughts?

Reply:
So now it's changed to skip when the message is null, can we do that? Because if auto complete is on it will be considered as "processed". Maybe deadlettering is better @wseye?

Reply:
Thoughts @wseye ?

Reply:
I'd suggest to merge this PR and we can still improve it via a dedicated PR.